rasa/core/featurizers/single_state_featurizer.py

```diff
@@ -59,12 +59,15 @@ def convert_to_dict(feature_states: List[Text]) -> Dict[Text, int]:
     def _state_features_for_attribute(
         self, sub_state: SubState, attribute: Text
     ) -> Dict[Text, int]:
+        # FIXME: the code below is not type-safe, but fixing it
+        #        would require more refactoring, for instance using
```

Comment:
what is the not type safe part about it?

Reply:
[`State`](https://github.com/RasaHQ/rasa/blob/b901d55907f9584dc4f37925e4f883c0fc7da9dc/rasa/shared/core/domain.py#L72e) is defined as a dictionary of `SubState`, which is defined as `Dict[Text, Union[Text, Tuple[Union[float, Text]]]]`, which is not actually true. The way the code uses the state: for a specific key of the `State` dict, the code expects a specific `SubState` structure. For instance:
- for `INTENT`, the `SubState` is `Dict[Text, Text]`
- for `ACTION_NAME`, the `SubState` is `TypedDict(intent=Text, text=Text, entities=List[Text])`
- for `SLOT`, the `SubState` is `Dict[Text, Union[Text, Tuple[float]]]`

`_state_features_for_attribute()` is kind of uncorrelating this dependency and assumes that because the attribute is of a certain value, the sub state will have a certain structure. Which can become risky. Does that make sense?

Reply:
Thanks, makes sense! Should we also fix the type annotation for `SubState` then? At least the `ACTION_NAME` case doesn't seem to be covered, is it?

Reply:
honestly I've tried and I couldn't make it work without refactoring the code in `rasa/core/featurizers/single_state_featurizer.py` and I was afraid of the consequences 😬 I don't know how much effort would be required to to it properly

Reply:
ok, let's skip it then